File: firmware/src/ir/denon.rs
use crate::ir::{
    utils::{decode_pulse_distance_data, match_mark, Endian},
    Protocol, MICROS_PER_TICK,
};

use super::IrSignal;

// Denon publish all their IR codes:
//  https://www.google.co.uk/search?q=DENON+MASTER+IR+Hex+Command+Sheet
//  -> http://assets.denon.com/documentmaster/us/denon%20master%20ir%20hex.xls

// MSB first, no start bit, 5 address + 8 command + 2 frame + 1 stop bit - each frame 2 times
//
const DENON_ADDRESS_BITS: usize = 5;
const DENON_COMMAND_BITS: usize = 8;
const DENON_FRAME_BITS: usize = 2; // 00/10 for 1. frame Denon/Sharp, inverted for autorepeat frame

const DENON_BITS: usize = DENON_ADDRESS_BITS + DENON_COMMAND_BITS + DENON_FRAME_BITS; // 15 - The number of bits in the command
const DENON_UNIT: u16 = 260;

const DENON_BIT_MARK: u16 = DENON_UNIT; // The length of a Bit:Mark
const DENON_ONE_SPACE: u16 = 7 * DENON_UNIT; // 1820 // The length of a Bit:Space for 1's
const DENON_ZERO_SPACE: u16 = 3 * DENON_UNIT; // 780 // The length of a Bit:Space for 0's

const DENON_AUTO_REPEAT_SPACE: u16 = 45000; // Every frame is auto repeated with a space period of 45 ms and the command inverted.
const _DENON_REPEAT_PERIOD: u32 = 110000; // Commands are repeated every 110 ms (measured from start to start) for as long as the key on the remote control is held down.

// for old decoder
const DENON_HEADER_MARK: u16 = DENON_UNIT; // The length of the Header:Mark
const _DENON_HEADER_SPACE: u16 = 3 * DENON_UNIT; // 780 // The length of the Header:Space

pub fn decode(signal: &[u16], last_decoded_command: u16) -> Option<IrSignal> {
    // we have no start bit, so check for the exact amount of data bits
    // Check we have the right amount of data (32). The + 2 is for initial gap + stop bit mark
    if signal.len() != (2 * DENON_BITS) + 2 {
        return Option::None;
    }

    // Read the bits in
    let decoded_data_opt = decode_pulse_distance_data(
        &signal[1..],
        DENON_BIT_MARK,
        DENON_ONE_SPACE,
        DENON_ZERO_SPACE,
        Endian::MSB,
    );
    let decoded_data = match decoded_data_opt {
        Option::None => return Option::None,
        Option::Some(v) => v,
    };

    // Check for stop mark
    if !match_mark(signal[(2 * DENON_BITS) + 1], DENON_HEADER_MARK) {
        return Option::None;
    }

    // Success
    let _endian = Endian::MSB;
    let frame_bits: u8 = (decoded_data & 0x03) as u8;
    let mut command: u16 = (decoded_data >> DENON_FRAME_BITS) as u16;
    let address: u16 = (command >> DENON_COMMAND_BITS) as u16;
    command = command & 0xff;
    let mut _repeat_count = 0;
    let mut _auto_repeat = false;
    let mut _parity_failed = false;
    let mut _repeat = false;

    // check for auto repeated inverted command
    // TODO verify MICROS_PER_TICK
    if signal[0] < ((DENON_AUTO_REPEAT_SPACE + (DENON_AUTO_REPEAT_SPACE / 4)) / MICROS_PER_TICK) {
        _repeat_count += 1;
        if frame_bits == 0x3 || frame_bits == 0x1 {
            // We are in the auto repeated frame with the inverted command
            _auto_repeat = true;
            // Check parity of consecutive received commands. There is no parity in one data set.
            let last_command = last_decoded_command;
            if last_command != !command {
                _parity_failed = true;
            }
            // always take non inverted command
            command = last_command;
        }
        if _repeat_count > 1 {
            _repeat = true;
        }
    } else {
        _repeat_count = 0;
    }

    let _number_of_bits = DENON_BITS;
    let protocol = if frame_bits == 1 || frame_bits == 2 {
        Protocol::Sharp
    } else {
        Protocol::Denon
    };
    return Option::Some(IrSignal {
        protocol,
        address,
        command,
    });
}

File: firmware/src/ir/Denon.cpp
#include "Denon.hpp"

namespace ir {

bool Denon::decode(DecoderState& state, DecoderResults* results) {
  // we have no start bit, so check for the exact amount of data bits
  // Check we have the right amount of data (32). The + 2 is for initial gap + stop bit mark
  if (state.buffer.getAvailable() != (2 * DENON_BITS) + 2) {
    return false;
  }

  // Read the bits in
  uint32_t tDecodedData;
  bool decoded = decodePulseDistanceData(
      state, DENON_BITS, 1, DENON_BIT_MARK, DENON_ONE_SPACE, DENON_ZERO_SPACE, Endian::MSB, &tDecodedData);
  if (!decoded) {
    return false;
  }

  // Check for stop mark
  if (!matchMark(state.buffer[(2 * DENON_BITS) + 1], DENON_HEADER_MARK)) {
    return false;
  }

  // Success
  results->endian = Endian::MSB;
  uint8_t tFrameBits = tDecodedData & 0x03;
  results->command = tDecodedData >> DENON_FRAME_BITS;
  results->address = results->command >> DENON_COMMAND_BITS;
  uint8_t tCommand = results->command & 0xFF;
  results->command = tCommand;

  // check for autorepeated inverted command
  if (state.buffer[0] < ((DENON_AUTO_REPEAT_SPACE + (DENON_AUTO_REPEAT_SPACE / 4)) / MICROS_PER_TICK)) {
    state.repeatCount++;
    if (tFrameBits == 0x3 || tFrameBits == 0x1) {
      // We are in the auto repeated frame with the inverted command
      results->autoRepeat = true;
      ;
      // Check parity of consecutive received commands. There is no parity in one data set.
      uint8_t tLastCommand = state.lastDecodedCommand;
      if (tLastCommand != (uint8_t)(~tCommand)) {
        results->parityFailed = true;
      }
      // always take non inverted command
      results->command = tLastCommand;
    }
    if (state.repeatCount > 1) {
      results->repeat = true;
    }
  } else {
    state.repeatCount = 0;
  }

  results->numberOfBits = DENON_BITS;
  if (tFrameBits == 1 || tFrameBits == 2) {
    results->protocol = Protocol::Sharp;
  } else {
    results->protocol = Protocol::Denon;
  }
  return true;
}

bool Denon::encode(
    peripheral::IrTx& irTx, Protocol protocol, uint32_t aAddress, uint32_t aCommand, uint32_t aNumberOfRepeats) const {
  irTx.reset(DENON_KHZ * 1000);

  // Shift command and add frame marker
  uint16_t tCommand = aCommand << DENON_FRAME_BITS;  // the lowest bits are 00 for Denon and 10 for Sharp
  if (protocol == Protocol::Sharp) {
    tCommand |= 0x02;
  }
  uint16_t tData = tCommand | ((uint16_t)aAddress << (DENON_COMMAND_BITS + DENON_FRAME_BITS));
  uint16_t tInvertedData = ((~tCommand) & 0x3FF) | (uint16_t)aAddress << (DENON_COMMAND_BITS + DENON_FRAME_BITS);

  uint_fast8_t tNumberOfCommands = aNumberOfRepeats + 1;
  while (tNumberOfCommands > 0) {
    irTx.reloadWatchdogCounter();

    // Data
    sendPulseDistanceWidthData(
        irTx,
        DENON_BIT_MARK,
        DENON_ONE_SPACE,
        DENON_BIT_MARK,
        DENON_ZERO_SPACE,
        tData,
        DENON_BITS,
        Endian::MSB,
        StopBit::True);
    irTx.waitForSendToComplete();

    // Inverted autorepeat frame
    irTx.delayMicros(DENON_AUTO_REPEAT_SPACE);
    sendPulseDistanceWidthData(
        irTx,
        DENON_BIT_MARK,
        DENON_ONE_SPACE,
        DENON_BIT_MARK,
        DENON_ZERO_SPACE,
        tInvertedData,
        DENON_BITS,
        Endian::MSB,
        StopBit::True);
    irTx.waitForSendToComplete();

    tNumberOfCommands--;
    // skip last delay!
    if (tNumberOfCommands > 0) {
      // send repeated command with a fixed space gap
      irTx.delayMicros(DENON_AUTO_REPEAT_SPACE);
    }
  }

  return true;
}

}  // namespace ir
